### PR TITLE
Onefile: Respect XDG_CACHE_HOME when rendering {CACHE_DIR} on non-Windows

### PR DESCRIPTION
**Pull request by xen0n**
# What does this PR do?

On non-Windows platforms, make `{CACHE_DIR}` expand to `$XDG_CACHE_HOME` if that variable is present in the environment. If it is absent the current behavior of using `~/.cache` is preserved.

# Why was it initiated? Any relevant Issues?

During work on https://github.com/ruyisdk/ruyi/issues/14 (issue description is in Chinese but the shell session should be evident enough) it was discovered the path pattern of `~/.cache` is currently hard-coded, while it is in fact an XDG directory and can be overridden with `$XDG_CACHE_HOME`.

# PR Checklist

- [x] Correct base branch selected? Should be `develop` branch.
- [x] Enabled commit hook or executed `./bin/autoformat-nuitka-source`.
- [x] All tests still pass. Check the Developer Manual about `Running the Tests`. There are GitHub
  Actions tests that cover the most important things however, and you are welcome to rely on those,
  but they might not cover enough.
- [ ] Ideally new features or fixed regressions ought to be covered via new tests.
- [ ] Ideally new or changed features have documentation updates.

### CODE REVIEW

File: nuitka/build/static_src/HelpersFilesystemPaths.c
Comment (kayhayen) [reply nuitka/build/static_src/HelpersFilesystemPaths.c:910]: The functions take care of the length checks themselves, and abort in case of overflows, so they are safe to use.

**Comment by kayhayen** [comment]
I am not sure `XDG_CACHE_HOME` is correct to use, according to https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html it could also be `XDG_RUNTIME_DIR` but I guess, that's not for persisting files, and with CACHE_DIR it should not be temporary file mode, it's sort of correct.

**Comment by kayhayen** [comment]
I will merge this into the next hotfix directly, it seems important to respect user preferences, and might even be required to make things work for some users at all.

File: nuitka/build/static_src/HelpersFilesystemPaths.c
Comment (kayhayen) [inline nuitka/build/static_src/HelpersFilesystemPaths.c:908]: This difference between append and expansion to a specific target makes things really hard to read. I think I will have an `appendExpandedTemplatePath` helper at least for `expandTemplatePath` to use itself, such that it's easier to get right.

**Review by kayhayen — APPROVED** [review]

File: nuitka/build/static_src/HelpersFilesystemPaths.c
Comment (kayhayen) [reply nuitka/build/static_src/HelpersFilesystemPaths.c:914]: I don't think we can prevent being influenced by these variables to write to arbitrary locations.

**Comment by xen0n** [comment]
> I am not sure `XDG_CACHE_HOME` is correct to use, according to https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html it could also be `XDG_RUNTIME_DIR` but I guess, that's not for persisting files, and with CACHE_DIR it should not be temporary file mode, it's sort of correct.

FYI, I chose it because the current default cache directory happens to coincide with it. My thought was that if `XDG_CACHE_HOME` wasn't appropriate, then the current default might not either.

**Comment by kayhayen** [comment]
I totally agree. But I think %CACHE_DIR% is kind of meant like a persistent one, and I think run directories are for clearing them out if used. This is surely only better.

**Comment by kayhayen** [comment]
This is hotfix/2.1.4 branch and has a "merges" annotation, I hope that will mark the PR as merged. Previously I would merge these to develop and later rebase them once i merge the hotfix, but that's ugly and problematic, but I am not sure of the syntax to use, so forgive me, if it doesn't work out.

**Comment by kayhayen** [comment]
It's also on factory now as well.

Thanks for your report, this is fixed on the factory branch, which is a development version under rapid development. You can try it out by going here: https://nuitka.net/doc/factory.html 

Feedback if this is working is very welcome, just please do not share plans of doing it, but rather confirmations or denials of it working.

**Comment by xen0n** [comment]
> It's also on factory now as well.
> 
> Thanks for your report, this is fixed on the factory branch, which is a development version under rapid development. You can try it out by going here: https://nuitka.net/doc/factory.html
> 
> Feedback if this is working is very welcome, just please do not share plans of doing it, but rather confirmations or denials of it working.

The change itself should be fine though: I have been building my project with the equivalent change [applied for ~4 months already](https://github.com/ruyisdk/ruyi/commits/main/scripts/patches/0001-Onefile-Respect-XDG_CACHE_HOME-when-rendering-CACHE_.patch), and our QA has confirmed that the original bug has been fixed. I only found time to push this upstream today though.

Please take your time to prepare the release; and thank you for handling this so quickly!

**Comment by xen0n** [comment]
> This is hotfix/2.1.4 branch and has a "merges" annotation, I hope that will mark the PR as merged. Previously I would merge these to develop and later rebase them once i merge the hotfix, but that's ugly and problematic, but I am not sure of the syntax to use, so forgive me, if it doesn't work out.

I think it's a simple check that if the HEAD of the PR is reachable from the PR's target branch, then the PR state would become "merged", but otherwise the PR should stay open. No problem though; feel free to just close it. Or you could force-push to my branch (I think I've not touched the default of allowing such maintainer pushes) so that the actual commit gets recognized by GitHub.

**Comment by kayhayen** [comment]
I have been told that similar to "fixes:" which closes issues, a "merges" can be used for PRs. That would be super cool. I understand you are a seasoned developer, but for others, declined PRs (i.e. not merged status) are an actual loss.

Thanks for your contribution. And upstream sooner. ;-)

**Comment by kayhayen** [comment]
Manually merged and part of 2.1.4 which I am releasing now.

**Pull request closed without merging**
